app/src/main/java/org/xbmc/kore/utils/Task.java

```diff
@@ -0,0 +1,81 @@
+package org.xbmc.kore.utils;
+
+import android.support.annotation.NonNull;
+
+/**
+ * Represents a computation that will eventually complete.
+ *
+ * @param <T> The type of the value being computed or awaited.
+ */
+public interface Task<T> {
+
+    /**
+     * Starts the computation, invokes a procedure when done.
+     *
+     * @param then The consumer of the value being computed.
+     */
+    void start(@NonNull OnFinish<? super T> then);
+
+    /**
+     * A procedure that awaits a value.
+     *
+     * @param <T> The type of the value generated by a previous computation.
+     */
+    interface OnFinish<T> {
+        void got(T result);
+    }
+
+    /**
+     * A sequence of tasks chained together.
+     *
+     * The values generated by the individual tasks are ignored except for the
+     * last one which may be accessed in the consumer if all of the tasks in
+     * the chain completed successfully.
+     *
+     * @param <T> The type of the value generated by the last task.
+     */
+    class Sequence<T> implements Task<T> {
```

Comment:
I wouldn't say it's coupled to Sequence. Ideally you shouldn't care about whether something is a Sequence or a Task. A Sequence is also a Task. You can pass it to `#then(Task)` or even do `new Sequence<>(sequence)`. It is possible to run a Task that's not in a sequence. I do exactly that in `#isPlayingVideo(handler).start(...)` in the fragment code. You can also write a named class that implements Task and use an instance of that in a Sequence.